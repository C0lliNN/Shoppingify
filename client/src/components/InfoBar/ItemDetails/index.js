import PropTypes from 'prop-types';
import React from 'react';
import BackButton from '../../BackButton';
import * as variables from '../../../helpers/style-constants';
import {
  showListBuilder,
  removeItem,
  addItemToList,
} from '../../../store/actions';
import { useDispatch } from 'react-redux';
import Button from '../../UI/Button/Button';
import ButtonBar from '../../ButtonBar';
import { useState } from 'react';
import getAxios from '../../../helpers/axios';
import Modal from '../../UI/Modal';
import Spinner from '../../UI/Spinner/Spinner';
import ErrorMessage from '../../ErrorMessage';
import {
  Container,
  ImageContainer,
  Title,
  Value,
  InfoGroupWrapper,
} from './styles';

function ItemDetails({ item }) {
  const [showModal, setShowModal] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState();

  const dispatch = useDispatch();

  async function handleDeleteItem() {
    const axios = getAxios();
    setIsLoading(true);
    try {
      await axios.delete(`/items/${item._id}`);

      setShowModal(false);
      setIsLoading(false);
      dispatch(removeItem(item));
      dispatch(showListBuilder());
    } catch (error) {
      if (error.response) {
        setError(error.response.data.message);
      } else {
        setError(error.message);
      }
      setIsLoading(false);
    }
  }

  function handleAddItemToList() {
    dispatch(addItemToList(item));
    dispatch(showListBuilder());
  }

  function closeModal() {
    setShowModal(false);
  }

  let modalContent = null;

  if (isLoading) {
    modalContent = <Spinner />;
  } else if (error) {
    modalContent = <ErrorMessage message={error} />;
  }

  const imgHolder = {};

  if (item.image) {
    imgHolder.backgroundImage = `url('${item.image}')`;
    imgHolder.backgroundSize = 'cover';
    imgHolder.backgroundPosition = 'center';
  }

  imgHolder.backgroundColor = variables.COLORS.gray_2;

  return (
    <Container>
      <BackButton onClick={() => dispatch(showListBuilder())} />
      <ImageContainer style={imgHolder} title={item.name} />
      <Title>name</Title>
      <Value style={{ fontSize: variables.FONT_SIZE_4 }}>{item.name}</Value>
      <Title>category</Title>
      <Value>{item.category.name}</Value>
      <InfoGroupWrapper>
        <Title>note</Title>
        <Value>{item.note ? item.note : 'No Notes'}</Value>
      </InfoGroupWrapper>
      <ButtonBar>
        <Button btnType="flat" onClick={() => setShowModal(true)}>
          delete
        </Button>
        <Button
          btnType="raised"
          variant="primary"
          onClick={handleAddItemToList}
        >
          Add to list
        </Button>
      </ButtonBar>
      {showModal && (
        <Modal
          title="Do you really want to remove this item?"
          onClose={closeModal}
          okButton={
            <Button
              btnType="raised"
              variant="danger"
              onClick={handleDeleteItem}
            >
              Yes
            </Button>
          }
          cancelButton={
            <Button btnType="flat" onClick={closeModal}>
              Cancel
            </Button>
          }
        >
          {modalContent}
        </Modal>
      )}
    </Container>
  );
}

ItemDetails.propTypes = {
  item: PropTypes.shape({
    _id: PropTypes.string,
    category: PropTypes.shape({
      name: PropTypes.string.isRequired,
    }).isRequired,
    image: PropTypes.string,
    name: PropTypes.string.isRequired,
    note: PropTypes.string,
  }),
};

export default ItemDetails;
