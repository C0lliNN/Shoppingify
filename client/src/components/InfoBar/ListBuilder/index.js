import React, { useEffect } from 'react';
import ButtonBar from '../../ButtonBar';
import Button from '../../UI/Button';
import bottle from '../../../assets/images/bottle.svg';
import { useDispatch, useSelector } from 'react-redux';
import {
  getActiveListHandler,
  saveListHandler,
  showCreateItem,
} from '../../../store/actions';
import { useState } from 'react';
import cartIllustration from '../../../assets/images/shopping.svg';
import SaveBar from './SaveBar/index.js';
import ItemsGroup from './ItemsGroup';
import {
  StyledListBuilder,
  Header,
  IllustrationContainer,
  BottleIllustration,
  Title,
  CustomButton,
  ListNameContainer,
  ListName,
  EditIcon,
  Text,
  CartIllustrationContainer,
  CartIllustration,
} from './styles';
import { completeListHandler } from '../../../store/actions';

function ListBuilder() {
  const [name, setName] = useState('');

  const list = useSelector((state) => state.activeList);
  const dispatch = useDispatch();

  const hasItems =
    list && list.data.filter((p) => p.category && p.category._id).length;

  function handleOnSave() {
    dispatch(saveListHandler(name));
  }

  function handleComplete() {
    dispatch(completeListHandler());
  }

  useEffect(() => {
    dispatch(getActiveListHandler());
  }, [dispatch]);

  return (
    <StyledListBuilder style={{ display: hasItems ? 'block' : 'flex' }}>
      <Header>
        <IllustrationContainer>
          <BottleIllustration src={bottle} alt="bottle illustration" />
          <div style={{ marginLeft: '5px' }}>
            <Title>{"Didn't find what you need?"}</Title>
            <CustomButton onClick={() => dispatch(showCreateItem())}>
              Add Item
            </CustomButton>
          </div>
        </IllustrationContainer>
        <ListNameContainer>
          <ListName>{list.name ? list.name : 'Shopping List'}</ListName>
          <EditIcon className="material-icons-round">edit</EditIcon>
        </ListNameContainer>
      </Header>
      {!hasItems && <Text>No Items</Text>}
      {!hasItems && (
        <CartIllustrationContainer>
          <CartIllustration src={cartIllustration} alt="" />
        </CartIllustrationContainer>
      )}
      {hasItems
        ? list.data
            .filter((p) => p.category && p.category._id)
            .sort((a, b) => a.category.name.localeCompare(b.category.name))
            .map((group) => (
              <ItemsGroup
                key={group.category._id}
                group={group}
                completing={list.saved}
              />
            ))
        : null}
      <ButtonBar>
        {list.saved ? (
          <>
            <Button btnType="flat">cancel</Button>
            <Button
              btnType="raised"
              variant="secondary"
              onClick={handleComplete}
            >
              complete
            </Button>
          </>
        ) : (
          <SaveBar
            disabled={!hasItems}
            onSave={handleOnSave}
            name={name}
            setName={setName}
          />
        )}
      </ButtonBar>
    </StyledListBuilder>
  );
}

export default ListBuilder;
