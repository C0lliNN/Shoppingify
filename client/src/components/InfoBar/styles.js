import styled from 'styled-components';
import * as variables from '../../helpers/style-constants';

export const StyledInfoBar = styled.aside`
  height: 100%;
  @media (min-width: ${variables.MD_BREAK_POINT}px) {
    position: fixed;
    width: ${variables.INFO_BAR_MD_SIZE}px;
    right: 0;
    top: 0;
  }
  @media (min-width: ${variables.LG_BREAK_POINT}px) {
    width: ${variables.INFO_BAR_LG_SIZE}px;
  }
`;
