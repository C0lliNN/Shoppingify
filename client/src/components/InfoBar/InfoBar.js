import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import * as variables from '../../helpers/style-constants';
import { connect } from 'react-redux';
import { SHOW_ITEM_DETAILS } from '../../store/actions/actionTypes';
import ItemDetails from './ItemDetails/ItemDetails';

const StyledInfoBar = styled.aside`
  height: 100%;
  @media (min-width: ${variables.MD_BREAK_POINT}px) {
    position: fixed;
    width: ${variables.INFO_BAR_MD_SIZE}px;
    right: 0;
    top: 0;
  }
  @media (min-width: ${variables.LG_BREAK_POINT}px) {
    width: ${variables.INFO_BAR_LG_SIZE}px;
  }
`;

function InfoBar({ contentType, item }) {
  let content = null;

  if (contentType === SHOW_ITEM_DETAILS) {
    content = <ItemDetails item={item} />;
  } else {
    content = <div>InfoBar</div>;
  }

  return <StyledInfoBar>{content}</StyledInfoBar>;
}

InfoBar.propTypes = {
  contentType: PropTypes.any,
  item: PropTypes.any,
};

const mapStateToProps = (state) => {
  return {
    contentType: state.infoBar.contentType,
    item: state.infoBar.item,
  };
};

export default connect(mapStateToProps)(InfoBar);
