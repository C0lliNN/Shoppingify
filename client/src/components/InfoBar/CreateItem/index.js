import React from 'react';
import ButtonBar from '../../ButtonBar';
import Button from '../../UI/Button';
import FormGroup from '../../UI/FormGroup';
import CategoryFormGroup from '../../CategoryFormGroup';
import { useState } from 'react';
import { showListBuilder } from '../../../store/actions';
import { useDispatch, useSelector } from 'react-redux';
import Modal from '../../UI/Modal';
import getAxios from '../../../helpers/axios';
import Spinner from '../../UI/Spinner/Spinner';
import { addItem } from '../../../store/actions';
import { StyledCreateItem, Title, ValidationError } from './styles';
import { useForm } from 'react-hook-form';

function CreateItem() {
  const [isLoading, setIsLoading] = useState(false);

  const [showModal, setShowModal] = useState(false);
  const [modalTitle, setModalTitle] = useState(null);

  const dispatch = useDispatch();
  const categories = useSelector((state) =>
    state.itemsData.data.map((group) => group.category)
  );
  const { register, handleSubmit, errors } = useForm();

  const nameRef = register({
    required: { value: true, message: 'The name is required' },
    minLength: {
      value: 2,
      message: 'The name must contain between 2 and 40 chars',
    },
    maxLength: {
      value: 40,
      message: 'The name must contain between 2 and 40 chars',
    },
  });

  const noteRef = register({
    maxLength: {
      value: 255,
      message: 'The note must contain at most 255 chars',
    },
  });

  const imageRef = register({
    pattern: {
      value: /^(http:\/\/www\.|https:\/\/www\.|http:\/\/|https:\/\/)?[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/i,
      message: 'The image must be a valid url',
    },
  });

  const categoryRef = register({
    required: { value: true, message: 'The category is required' },
    minLength: {
      value: 4,
      message: 'The category must contain between 4 and 120 chars',
    },
    maxLength: {
      value: 120,
      message: 'The category must contain between 4 and 120 chars',
    },
  });

  async function onSubmit(data, event) {
    event.preventDefault();

    const { name, note, category, image } = data;

    let categoryPayload = categories.find(
      (categoryItem) => categoryItem.name === category
    );

    if (!categoryPayload) {
      categoryPayload = { name: category };
    }

    const payload = {
      name,
      note,
      image,
      category: categoryPayload,
    };

    try {
      const response = await getAxios().post('/items', payload);
      dispatch(addItem(response.data));

      dispatch(showListBuilder());
    } catch (error) {
      if (error.response) {
        setModalTitle(error.response.data.message);
      } else {
        setModalTitle(error.message);
      }
      setShowModal(true);
    } finally {
      setIsLoading(false);
    }
  }

  function closeModal() {
    setShowModal(false);
  }

  console.log(errors);

  return (
    <StyledCreateItem>
      <form onSubmit={handleSubmit(onSubmit)} action="#" method="post">
        <Title>Add a new item</Title>
        {isLoading ? (
          <Spinner />
        ) : (
          <React.Fragment>
            <FormGroup>
              <FormGroup.Label htmlFor="name">Name</FormGroup.Label>
              <FormGroup.Input
                placeholder="Enter a name"
                ref={nameRef}
                id="name"
                name="name"
              />
              {errors?.name && (
                <ValidationError>{errors?.name.message}</ValidationError>
              )}
            </FormGroup>

            <FormGroup>
              <FormGroup.Label htmlFor="note">Note(optional)</FormGroup.Label>
              <FormGroup.Textarea
                placeholder="Enter a note"
                id="note"
                rows="3"
                name="note"
                ref={noteRef}
              ></FormGroup.Textarea>
              {errors?.note && (
                <ValidationError>{errors?.note.message}</ValidationError>
              )}
            </FormGroup>
            <FormGroup>
              <FormGroup.Label htmlFor="image">Image(optional)</FormGroup.Label>
              <FormGroup.Input
                placeholder="Enter a url"
                name="image"
                id="image"
                ref={imageRef}
              />
              {errors?.image && (
                <ValidationError>{errors?.image.message}</ValidationError>
              )}
            </FormGroup>
            <CategoryFormGroup
              categories={categories}
              ref={categoryRef}
              errors={errors}
            />
          </React.Fragment>
        )}
        <ButtonBar>
          <Button
            btnType="flat"
            onClick={() => dispatch(showListBuilder())}
            type="button"
          >
            cancel
          </Button>
          <Button btnType="raised" variant="primary" type="submit">
            Save
          </Button>
        </ButtonBar>
      </form>
      {showModal && (
        <Modal
          title={modalTitle}
          onClose={closeModal}
          okButton={
            <Button btnType="raised" variant="danger" onClick={closeModal}>
              OK
            </Button>
          }
        />
      )}
    </StyledCreateItem>
  );
}

export default CreateItem;
