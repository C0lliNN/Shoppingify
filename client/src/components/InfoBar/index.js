import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import {
  SHOW_ITEM_DETAILS,
  SHOW_CREATE_ITEM,
} from '../../store/actions/actionTypes';
import ItemDetails from './ItemDetails';
import ListBuilder from './ListBuilder/index.js';
import CreateItem from './CreateItem';
import { StyledInfoBar } from './styles';


function InfoBar({ contentType, item }) {
  let content = null;

  if (contentType === SHOW_ITEM_DETAILS) {
    content = <ItemDetails item={item} />;
  } else if (contentType === SHOW_CREATE_ITEM) {
    content = <CreateItem />;
  } else {
    content = <ListBuilder />;
  }

  return <StyledInfoBar>{content}</StyledInfoBar>;
}

InfoBar.propTypes = {
  contentType: PropTypes.any,
  item: PropTypes.any,
};

const mapStateToProps = (state) => {
  return {
    contentType: state.infoBar.contentType,
    item: state.infoBar.item,
  };
};

export default connect(mapStateToProps)(InfoBar);
