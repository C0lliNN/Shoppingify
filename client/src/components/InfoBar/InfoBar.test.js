import React from 'react';
import { render } from '../../tests/utilities';
import InfoBar from './InfoBar';
import 'jest-styled-components';

describe('<InfoBar/>', () => {
  it('should render correctly', () => {
    const { container } = render(<InfoBar />);
    expect(container).toMatchSnapshot();
  });
});
