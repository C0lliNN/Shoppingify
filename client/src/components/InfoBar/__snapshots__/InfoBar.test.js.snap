// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoBar/> should render correctly 1`] = `
<div>
  <aside
    class="sc-fzpmMD dPlWFi"
  >
    <div
      class="sc-AxmLO ebtaiT"
    >
      <div
        class="sc-fzozJi gOqWKZ"
      >
        <img
          alt="bottle illustration"
          class="sc-fzpans bNRjHC"
          src="bottle.svg"
        />
        <div
          style="margin-left: 5px;"
        >
          <h3
            class="sc-fzoLsD jEAnWv"
          >
            Didn't find what you need?
          </h3>
          <button
            class="sc-fzplWN chnRYW"
          >
            Add Item
          </button>
        </div>
      </div>
      <div
        class="sc-fznyAO kdKWyA"
      >
        <h3
          class="sc-fznKkj eLkAGK"
        >
          Shopping List
        </h3>
        <i
          class="sc-fznZeY bppYyw material-icons-round"
        >
          edit
        </i>
      </div>
      <div
        class="sc-AxhCb ohEyP"
      >
        <button
          class="sc-AxiKw VDLGN"
        >
          cancel
        </button>
        <button
          class="sc-AxiKw hSJBCn"
        >
          complete
        </button>
      </div>
    </div>
  </aside>
</div>
`;
